File: compiler/src/lexer.rs
use std::fmt;

use token::{Token, TokenType, Position};


pub struct Lexer {
	source: String,
	tokens: Vec<Token>,
}
impl Lexer {
	pub fn new(source: String) -> Lexer {
		let mut tokens = Vec::new();

		let mut char_iter = source.chars();
		let mut next_char: Option<char> = char_iter.next();

		let mut current_string = String::new();

		let mut position = Position::new(0, 1, 1);
		let mut previous_position = Position::new(0, 1, 1);

		macro_rules! push_c {
			($c:expr) => (
				current_string.push($c);
				position.add($c);
				next_char = char_iter.next();
			)
		}
		while let Some(c) = next_char {
			position.add(c);
			next_char = char_iter.next(); //cut this, perhaps make some helper to eat / not?
			let mut token_type = TokenType::Unknown;
			match c { 
				' ' | '\t' | '\n' | '\r' => {
					//while let Some(c @ (' ' | '\t' | '\n' | '\r')) = next_char { position.add(c); }
					token_type = TokenType::Whitespace;
				},
				'0'..='9' => {
					current_string.push(c);
					while let Some(c @ '0'..='9') | Some(c @ '.') = next_char { push_c!(c); } //will need to add . and possible other descriptors
					if current_string.contains('.') {
						token_type = TokenType::LitFloat(current_string.parse().unwrap());
					} else {
						token_type = TokenType::LitInt(current_string.parse().unwrap());
					}
				}
				'a'..='z' | 'A'..='Z' | '_' => {
					current_string.push(c);
					let first_char = c;
					let mut number_count = 0;
					while let Some(c @ ('a'..='z' | 'A'..='Z' | '_' | '0'..='9')) = next_char {
						if c.is_digit(10) {
							number_count += 1;
						}
						push_c!(c);
					}
					if current_string.len() - number_count == 1 && number_count != 0  {
						match first_char {
							'u' => token_type = TokenType::UIntType(current_string[1..].parse().unwrap()),
							'i' => token_type = TokenType::IntType(current_string[1..].parse().unwrap()),
							'f' => token_type = TokenType::FloatType(current_string[1..].parse().unwrap()),
							_ => token_type = TokenType::Ident(current_string.clone()),
						}
					} else {
						match current_string.as_str() {
							"type" => token_type = TokenType::Type, 
							"fn" => token_type = TokenType::Fn,
							"let" => token_type = TokenType::Let,
							"if" => token_type = TokenType::If,
							"match" => token_type = TokenType::Match,
							"as" => token_type = TokenType::As,
							"to" => token_type = TokenType::To,
							"is" => token_type = TokenType::Is,
							"true" => token_type = TokenType::True,
							"false" => token_type = TokenType::False,
							"tail" => token_type = TokenType::Tail,
							"rec" => token_type = TokenType::Rec,
							"bool" => token_type = TokenType::BoolType,
							_ => token_type = TokenType::Ident(current_string.clone()), 
						}
					}
				}
				'.' => {
					current_string.push(c);
					while let Some(c @ '0'..='9') = next_char { push_c!(c); }
					if current_string.len() == 1 {
						token_type = TokenType::Dot;
					} else {
						token_type = TokenType::LitFloat(current_string.parse().unwrap());
					}
				}
				'!' => token_type = TokenType::Bang,
				'/' => token_type = TokenType::FSlash,
				'\\' => token_type = TokenType::BSlash,
				'*' => token_type = TokenType::Star,
				'%' => token_type = TokenType::Mod,
				'&' => {
					next_char = char_iter.next();
					match next_char {
						Some('&') => {
							token_type = TokenType::And;
							position.add('&');
						},
						_ => token_type = TokenType::Amp,
					}
				},
				'{' => token_type = TokenType::LCurly, 
				'}' => token_type = TokenType::RCurly,
				'(' => token_type = TokenType::LParen,
				')' => token_type = TokenType::RParen,
				'[' => token_type = TokenType::LSquare,
				']' => token_type = TokenType::RSquare,
				'<' => {
					next_char = char_iter.next();
					match next_char {
						Some('=') => {
							token_type = TokenType::Le;
							position.add('=');
						},
						/*Some('>') => {
							token_type = TokenType::Ne;
							position.add('>');
						},*/
						Some('|') => {
							token_type = TokenType::LPipe;
							position.add('|');
						},
						Some('-') => {
							token_type = TokenType::LArrow;
							position.add('-');
						},
						_ => token_type = TokenType::Lt,
					}
				},
				'>' => {
					next_char = char_iter.next();
					match next_char {
						Some('=') => {
							token_type = TokenType::Ge;
							position.add('=');
						},
						/*Some('>') => {
							token_type = TokenType::Ne;
							position.add('>');
						},*/
						_ => token_type = TokenType::Gt,
					}
				},
				'|' => {
					next_char = char_iter.next();
					match next_char {
						Some('|') => {
							token_type = TokenType::Or;
							position.add('|');
						},
						Some('>') => {
							token_type = TokenType::RPipe;
							position.add('>');
						},
						_ => token_type = TokenType::Bar,
					}
				},
				'^' => token_type = TokenType::Xor,
				'+' => token_type = TokenType::Plus,
				'-' => {
					next_char = char_iter.next();
					match next_char {
						Some('>') => {
							token_type = TokenType::RArrow;
							position.add('>');
						},
						_ => token_type = TokenType::Minus,
					}
				},
				'=' => {
					next_char = char_iter.next();
					match next_char {
						Some('=') => {
							token_type = TokenType::Eq;
							position.add('=');
						},
						_ => token_type = TokenType::Assign,
					}
				},
				':' => token_type = TokenType::Colon,
				';' => token_type = TokenType::Semi,
				_ => { }
			}
			tokens.push(Token::new(token_type, previous_position.clone()));
			previous_position = position.clone();
			current_string.clear();
		}
		Lexer {
			source,
			tokens,
		}
	}
}
impl fmt::Display for Lexer {
	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
		for token in &self.tokens {
			write!(f, "{}\n", token)?;
		}
		Ok(())
	}	
}
impl Lexer {
	pub fn tokens(&self) -> &Vec<Token> {
		&self.tokens
	}
	pub fn source(&self) -> &String {
		&self.source
	}
}

File: compilerex/lib/lexer.ex
defmodule Birch.Lexer do
  defstruct tokens: [], input: ""

  defmodule Position do 
    defstruct index: 0, line: 0, column: 0
    def increment(position, char) do
      case char do
        c when c in [?\n, ?\r] -> %Position{index: position.index + 1, line: position.line + 1, column: 0}
        _ -> %Position{index: position.index + 1, line: position.line, column: position.column + 1}
      end
    end
  end

  defmodule Token do
    defstruct token: {:unknown}, position: %Position{}
  end

  defmacro is_alphabet(char) do
    quote do
      unquote(char) in ?a..?z or unquote(char) in ?A..?Z
    end
  end
  defmacro is_alphanum_underscore(char) do
    quote do
      unquote(char) in ?0..?9 or unquote(char) in ?a..?z or unquote(char) in ?A..?Z or unquote(char) == ?_
    end
  end
  @whitespace [?\s, ?\t, ?\r, ?\n]
  @numbers ?0..?9

  def tokenize(chars, position) do
    case chars do
      [] -> []
      [char | rest] -> <<char::utf8>> = char
        start_position = Position.increment(position, char)
          {token, final_position, rest} = case char do
            c when c in @whitespace -> {new_position, rest} = tokenize_whitespace(rest, start_position)
              {%Token{token: :whitespace, position: position}, new_position, rest}
            c when c in @numbers  -> {num, new_position, rest} = tokenize_number(rest, <<c::utf8>>,  start_position)
              {%Token{token: {:int, num}, position: position}, new_position, rest}
              #add more options, as well as perhaps the size
            c when is_alphabet(c) -> {size, new_position, rest} = tokenize_number(rest, <<c::utf8>>, start_position)
              if new_position != start_position do
                case c do
                  ?i -> {%Token{token: {:int_type, size}, position: position}, new_position, rest}
                  ?u -> {%Token{token: {:uint_type, size}, position: position}, new_position, rest}
                  ?f -> {%Token{token: {:float_type, size}, position: position}, new_position, rest}
                  _ -> {%Token{token: {:identifier, <<c::utf8>> <> size}, position: position}, new_position, rest}
                end
              else
                {identifier, new_position, rest} = tokenize_identifier(rest, <<c::utf8>>, start_position)
                case identifier do
                  "true" -> {%Token{token: :true, position: position}, new_position, rest}
                  "false" -> {%Token{token: :false, position: position}, new_position, rest}
                  "if" -> {%Token{token: :if, position: position}, new_position, rest}
                  "then" -> {%Token{token: :then, position: position}, new_position, rest}
                  "else" -> {%Token{token: :else, position: position}, new_position, rest}
                  "bool" -> {%Token{token: :bool_type, position: position}, new_position, rest}
                  "fn" -> {%Token{token: :fn, position: position}, new_position, rest}
                  "let" -> {%Token{token: :let, position: position}, new_position, rest}
                  "to" -> {%Token{token: :to, position: position}, new_position, rest}
                  "as" -> {%Token{token: :as, position: position}, new_position, rest}
                  "in" -> {%Token{token: :in, position: position}, new_position, rest}
                  "match" -> {%Token{token: :match, position: position}, new_position, rest}
                  "tail" -> {%Token{token: :tail, position: position}, new_position, rest}
                  "rec" -> {%Token{token: :rec, position: position}, new_position, rest}
                  "type" -> {%Token{token: :type, position: position}, new_position, rest}
                  "self" -> {%Token{token: :self, position: position}, new_position, rest}
                  _ -> {%Token{token: {:identifier, identifier}, position: position}, new_position, rest}
                end
              end
            ?* -> {%Token{token: :star, position: position}, start_position, rest}
            ?/ -> {%Token{token: :f_slash, position: position}, start_position, rest}
            ?+ -> {%Token{token: :plus, position: position}, start_position, rest}
            ?- -> new_position = Position.increment(start_position, ?-)
              case rest do
                [?> | rest] -> {%Token{token: :r_arrow, position: position}, new_position, rest}
                _ -> {%Token{token: :minus, position: position}, start_position, rest}
            end
            ?= -> new_position = Position.increment(start_position, ?=)
              case rest do
                [?> | rest] -> {%Token{token: :r_fat_arrow, position: position}, new_position, rest}
                [?= | rest] -> {%Token{token: :eq, position: position}, new_position, rest}
                _ -> {%Token{token: :eq, position: position}, start_position, rest}
              end
            ?| -> new_position = Position.increment(start_position, ?|)
              case rest do
                [?> | rest] -> {%Token{token: :r_point, position: position}, new_position, rest}
                [?| | rest] -> {%Token{token: :or, position: position}, new_position, rest}
                _ -> {%Token{token: :pipe, position: position}, start_position, rest}
              end
            ?& -> new_position = Position.increment(start_position, ?&)
              case rest do
                [?& | rest] -> {%Token{token: :and, position: position}, new_position, rest}
                _ -> {%Token{token: :amp, position: position}, start_position, rest}
              end
            ?! -> new_position = Position.increment(start_position, ?!)
              case rest do
                [?= | rest] -> {%Token{token: :neq, position: position}, new_position, rest}
                _ -> {%Token{token: :bang, position: position}, start_position, rest}
              end 
            ?< -> new_position = Position.increment(start_position, ?<)
              case rest do
                [?- | rest] -> {%Token{token: :l_arrow, position: position}, new_position, rest}
                [?| | rest] -> {%Token{token: :l_point, position: position}, new_position, rest}
                [?= | rest] -> {%Token{token: :leq, position: position}, new_position, rest}
                _ -> {%Token{token: :lt, position: position}, start_position, rest}
              end
            ?( -> {%Token{token: :l_paren, position: position}, start_position, rest}
            ?) -> {%Token{token: :r_paren, position: position}, start_position, rest}
            ?{ -> {%Token{token: :l_brace, position: position}, start_position, rest}
            ?} -> {%Token{token: :r_brace, position: position}, start_position, rest}
            ?[ -> {%Token{token: :l_square, position: position}, start_position, rest}
            ?] -> {%Token{token: :r_square, position: position}, start_position, rest}
            ?; -> {%Token{token: :semicolon, position: position}, start_position, rest}
            ?: -> {%Token{token: :colon, position: position}, start_position, rest}
            ?? -> {%Token{token: :q_mark, position: position}, start_position, rest}
            ?\\ -> {%Token{token: :b_slash, position: position}, start_position, rest}
            _ -> {%Token{token: {:unknown, char}, position: position}, start_position, rest}
          end
          [token | tokenize(rest, final_position)]
    end
  end
  def tokenize_identifier(chars, identifier_str, position) do
    case chars do
      [] -> {identifier_str, position, []}
      [char | rest] -> <<char::utf8>> = char
        case char do
          c when is_alphanum_underscore(c) -> tokenize_identifier(rest, identifier_str <> <<c::utf8>>, Position.increment(position, char))
          _ -> {identifier_str, position, chars}
        end
    end
  end
  def tokenize_number(chars, number_str, position) do
    case chars do
      [] -> {number_str, position, []}
      [char | rest] -> <<char::utf8>> = char
        case char do
          c when c in @numbers -> tokenize_number(rest, number_str <> <<c::utf8>>, Position.increment(position, char))
          _ -> {number_str, position, chars}
        end
    end
  end
  def tokenize_whitespace(chars, position) do
    case chars do
      [] -> {position, []}
      [char | rest] -> case char do
        c when c in @whitespace -> tokenize_whitespace(rest, Position.increment(position, char))
        _ -> {position, chars}
      end
    end
  end

end

